Framework/MDAlgorithms/inc/MantidMDAlgorithms/ConvToMDEventsWSIndexing.h

```diff
@@ -0,0 +1,530 @@
+          DataObjects::MDEventWorkspace<DataObjects::MDLeanEvent<ND>, ND> *>(
+          m_OutWSWrapper->pWorkspace().get())
+      )
+    return LEAN;
+  return NONE;
+}
+
+
+template<size_t ND, template <size_t> class MDEventType, typename EventIterator>
+void ConvToMDEventsWSIndexing::EventsDistributor<ND, MDEventType, EventIterator>::
+    distributeEvents(Task& tsk, const WORKER_TYPE& wtp) {
+  using MDEvent = MDEventType<ND>;
+  using MortonT = typename MDEvent::MortonT;
+  using BoxBase = DataObjects::MDBoxBase<MDEvent, ND>;
+  using Box = DataObjects::MDBox<MDEvent, ND>;
+  using GridBox = DataObjects::MDGridBox<MDEvent, ND>;
+
+
+  const size_t childBoxCount = tsk.bc->getNumSplit();
+  const size_t splitThreshold = tsk.bc->getSplitThreshold();
+
+  if (tsk.maxDepth-- == 1 || std::distance(tsk.begin, tsk.end) <= static_cast<int>(splitThreshold)) {
+    return;
+  }
+
+/* Determine the "width" of this box in Morton number */
+  const MortonT thisBoxWidth = tsk.upperBound - tsk.lowerBound;
+
+  /* Determine the "width" of the child boxes in Morton number */
+  const MortonT childBoxWidth = thisBoxWidth / childBoxCount;
+
+  auto eventIt = tsk.begin;
+
+  struct RecursionHelper {
+    std::pair<EventIterator, EventIterator> eventRange;
+    std::pair<MortonT, MortonT> mortonBounds;
+    BoxBase* box;
+  };
+  std::vector<RecursionHelper> children;
+  children.reserve(childBoxCount);
+
+  /* For each new child box */
+  for (size_t i = 0; i < childBoxCount; ++i) {
+    /* Lower child box bound is parent box lower bound plus for each previous
+     * child box; box width plus offset by one (such that lower bound of box
+     * i+1 is one grater than upper bound of box i) */
+    const auto boxLower = tsk.lowerBound + ((childBoxWidth + 1) * i);
+
+    /* Upper child box bound is lower plus child box width */
+    const auto boxUpper = childBoxWidth + boxLower;
+
+    const auto boxEventStart = eventIt;
+
+    if (md_structure_ws::morton_contains<MortonT>(boxLower, boxUpper,eventIt->getIndex()))
+      eventIt = std::upper_bound(boxEventStart, tsk.end, boxUpper,
+          [](const MortonT& m, const typename std::iterator_traits<EventIterator>::value_type& event){
+            return m < event.getIndex();
+          });
+
+    /* Add new child box. */
+    /* As we are adding as we iterate over Morton numbers and parent events
+     * child boxes are inserted in the correct sorted order. */
+    std::vector<Mantid::Geometry::MDDimensionExtents<coord_t>> extents(ND);
+    auto minCoord = MDEvent::indexToCoordinates(boxLower, tsk.space);
+    auto maxCoord = MDEvent::indexToCoordinates(boxUpper, tsk.space);
+    for(unsigned ax = 0; ax < ND; ++ax) {
+      extents[ax].setExtents(minCoord[ax], maxCoord[ax]);
+    }
+
+    BoxBase* newBox;
+    if(std::distance(boxEventStart, eventIt) <= static_cast<int>(splitThreshold) || tsk.maxDepth == 1) {
+      for(auto it = boxEventStart; it < eventIt; ++it)
+        it->retrieveCoordinates(tsk.space);
```

Comment:
It is done like this because I want to change MDBox class as less, as It possible: I don't want to either create a function that shares non const version of events in the box or provide the method to retrieve coordinates to the box. I want to leave all the conversion stuff inside the one class.

Reply:
But didn't you add a new constructor to it? The constructor could do the conversion.

Reply:
If the constructor does the conversion it increases the requirements to the EventType template argument, it should provide retrieveCoordinates method in this case. It is not a very good design in my point of view.